crates/miden-lib/asm/kernels/transaction/lib/faucet.masm

```diff
@@ -201,31 +197,66 @@ proc.burn_non_fungible_asset
     # => [ASSET, ASSET]
 
     # fetch the root of the SMT containing the non-fungible assets
-    exec.account::get_faucet_storage_data_slot exec.account::get_item
-    # => [SMT_ROOT, ASSET, ASSET]
+    exec.account::get_faucet_storage_data_slot dup movdn.5 exec.account::get_item
+    # => [SMT_ROOT, ASSET, faucet_storage_data_slot, ASSET]
 
     # prepare stack for removal of non-fungible asset from tracking SMT
-    swapw padw
-    # => [EMPTY_WORD, ASSET, SMT_ROOT, ASSET]
+    swapw exec.asset_vault::build_non_fungible_asset_vault_key padw swapw movup.12
+    # => [faucet_storage_data_slot, ASSET_KEY, EMPTY_WORD, SMT_ROOT, ASSET]
 
     # remove the non-fungible asset from the tracking SMT
-    exec.smt::set
-    # => [OLD_VAL, SMT_ROOT', ASSET]
-
-    # assert the `OLD_VAL` is not EMPTY_WORD, indicating that the non-fungible asset exists.
-    # we only need to check ASSET[1] as this is always set to the faucet_id and can not be 0.
-    drop drop eq.0 not assert.err=ERR_FAUCET_NON_FUNGIBLE_ASSET_TO_BURN_NOT_FOUND drop
-    # => [SMT_ROOT', ASSET]
+    exec.account::set_map_item dropw
+    # => [OLD_VAL, ASSET]
```

Comment:
Not from this PR, but why does `set_map_item` require us to pass `SMT_ROOT` to it? I wonder if that's something we could do inside `set_map_item`.

Reply:
I agree, I think it can be optimized. Created an issue: #1137